src/scib_metrics/_ari_nmi.py

```diff
@@ -33,9 +46,72 @@ def nmi_ari_cluster_labels(X: np.ndarray, labels: np.ndarray) -> Tuple[float, fl
     ari
         Adjusted rand index score
     """
+    X = check_array(X, accept_sparse=False, ensure_2d=True)
+    labels = check_array(labels, accept_sparse=False, ensure_2d=False)
     n_clusters = len(np.unique(labels))
-    labels_pred = _compute_clustering(X, n_clusters)
+    labels_pred = _compute_clustering_kmeans(X, n_clusters)
     nmi = normalized_mutual_info_score(labels, labels_pred, average_method="arithmetic")
     ari = adjusted_rand_score(labels, labels_pred)
 
     return nmi, ari
+
+
+def nmi_ari_cluster_labels_leiden(
+    X: spmatrix, labels: np.ndarray, optimize_resolution: bool = True, resolution: float = 1.0, n_jobs: int = 1
+) -> Tuple[float, float]:
+    """Compute nmi and ari between leiden clusters and labels.
+
+    This deviates from the original implementation in scib by using leiden instead of
+    louvain clustering. Installing joblib allows for parallelization of the leiden
+    resoution optimization.
+
+    Parameters
+    ----------
+    X
+        Array of shape (n_samples, n_samples) representing a connectivity graph.
+        Values should represent weights between pairs of neighbors, with a higher weight
+        indicating more connected.
+    labels
+        Array of shape (n_samples,) representing label values
+    optimize_resolution
+        Whether to optimize the resolution parameter of leiden clustering by searching over
+        10 values
+    resolution
+        Resolution parameter of leiden clustering. Only used if optimize_resolution is False.
+    n_jobs
+        Number of jobs for parallelizing resolution optimization via joblib. If -1, all CPUs
+        are used.
+
+    Returns
+    -------
+    nmi
+        Normalized mutual information score
+    ari
+        Adjusted rand index score
+    """
+    X = check_array(X, accept_sparse=False, ensure_2d=True)
+    if X.shape[0] != X.shape[1]:
+        raise ValueError("X should be a square matrix")
+    labels = check_array(labels, accept_sparse=False, ensure_2d=False)
+    if optimize_resolution:
+        n = 10
+        resolutions = np.array([2 * x / n for x in range(1, n + 1)])
+        try:
+            from joblib import Parallel, delayed
+
+            out = Parallel(n_jobs=n_jobs)(
+                delayed(nmi_ari_cluster_labels_leiden)(X, labels, False, r) for r in resolutions
+            )
+        except ImportError:
+            logger.info("Using for loop over resolutions. pip install joblib for parallelization.")
+            out = [nmi_ari_cluster_labels_leiden(X, labels, False, r) for r in resolutions]
```

Comment:
this recursive pattern feels like a recipe for bugs. would be much cleaner to just split out the else case into a helper then call that from both cases (optimize or not optimize)

Reply:
Done.

Reply:
speedy